test/libevmasm/PlainAssemblyParser.cpp

```diff
@@ -0,0 +1,182 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <test/libevmasm/PlainAssemblyParser.h>
+
+#include <test/Common.h>
+#include <test/libsolidity/util/SoltestErrors.h>
+
+#include <libevmasm/Instruction.h>
+
+#include <liblangutil/Common.h>
+
+#include <boost/algorithm/string/find.hpp>
+
+#include <fmt/format.h>
+
+#include <sstream>
+
+using namespace std::string_literals;
+using namespace solidity;
+using namespace solidity::test;
+using namespace solidity::evmasm;
+using namespace solidity::evmasm::test;
+using namespace solidity::langutil;
+
+Json PlainAssemblyParser::parse(std::string _sourceName, std::string const& _source)
+{
+	m_sourceName = std::move(_sourceName);
+	Json codeJSON = Json::array();
+	std::istringstream sourceStream(_source);
+	while (getline(sourceStream, m_line))
+	{
+		advanceLine(m_line);
+		if (m_lineTokens.empty())
+			continue;
+
+		if (c_instructions.contains(currentToken().value))
+		{
+			expectNoMoreArguments();
+			codeJSON.push_back({{"name", currentToken().value}});
+		}
+		else if (currentToken().value == "PUSH")
+		{
+			if (hasMoreTokens() && nextToken().value == "[tag]")
+			{
+				advanceToken();
+				std::string_view tagID = expectArgument();
+				expectNoMoreArguments();
+				codeJSON.push_back({{"name", "PUSH [tag]"}, {"value", tagID}});
+			}
+			else
+			{
+				std::string_view immediateArgument = expectArgument();
+				expectNoMoreArguments();
+
+				if (!immediateArgument.starts_with("0x"))
+					BOOST_THROW_EXCEPTION(std::runtime_error(formatError("The immediate argument to PUSH must be a hex number prefixed with '0x'.")));
+
+				immediateArgument.remove_prefix("0x"s.size());
+				codeJSON.push_back({{"name", "PUSH"}, {"value", immediateArgument}});
```

Comment:
It's quite annoying that the JSON format accepts only hex numbers here but without the `0x` prefix. On the other hand for `tag` it accepts both both and the ones without `0x` are actually decimal.